cpp/include/cudf/ast/detail/transform.cuh

```diff
@@ -318,41 +319,60 @@ __device__ void evaluate_row_expression(detail::row_evaluator const& evaluator,
 
 struct ast_plan {
  public:
-  ast_plan() : sizes(), data_pointers() {}
+  ast_plan(linearizer const& expr_linearizer) : _sizes{}, _data_pointers{}
+  {
+    add_to_plan(expr_linearizer.data_references());
+    add_to_plan(expr_linearizer.literals());
+    add_to_plan(expr_linearizer.operators());
+    add_to_plan(expr_linearizer.operator_source_indices());
+  }
 
   using buffer_type = std::pair<std::unique_ptr<char[]>, int>;
 
+  /**
+   * @brief Helper function for adding components (operators, literals, etc) to AST plan
+   *
+   * @tparam T  The underlying type of the input `std::vector`
+   * @param  v  The `std::vector` containing components (operators, literals, etc)
+   */
   template <typename T>
   void add_to_plan(std::vector<T> const& v)
   {
     auto const data_size = sizeof(T) * v.size();
-    sizes.push_back(data_size);
-    data_pointers.push_back(v.data());
+    _sizes.push_back(data_size);
+    _data_pointers.push_back(v.data());
   }
 
-  buffer_type get_host_data_buffer() const
+  /**
+   * @brief Create and return host buffer
+   *
+   * @return `std::pair` containing host buffer and buffer size
+   */
+  buffer_type host_data_buffer() const
   {
-    auto const total_size = std::accumulate(sizes.cbegin(), sizes.cend(), 0);
+    auto const total_size = std::accumulate(_sizes.cbegin(), _sizes.cend(), 0);
     auto host_data_buffer = std::make_unique<char[]>(total_size);
-    auto const offsets    = get_offsets();
-    for (unsigned int i = 0; i < data_pointers.size(); ++i) {
-      std::memcpy(host_data_buffer.get() + offsets[i], data_pointers[i], sizes[i]);
-    }
+    for (unsigned int i = 0; i < _data_pointers.size(); ++i)
+      std::memcpy(host_data_buffer.get() + offsets()[i], _data_pointers[i], _sizes[i]);
```

Comment:
Hm, the call to `offsets()` in the loop here gives me pause. Without any optimization, this is going to compute the scan for every iteration of the loop. I'm not 100% confident that the compiler will be able to make that optimization. I think you either need to put the call outside the loop or memoize the results from `offsets()`.

Reply:
Thats a good point.

Reply:
Alternatively, there isn't even a need for a public `offsets()` function if you move the `device_data_buffer` aliasing into the constructor too.

Reply:
Why not calculate offset in the for-loop? to allow loop unrolling by compiler?